File: presto-bigquery/src/main/java/com/facebook/presto/plugin/bigquery/Conversions.java
/*
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.facebook.presto.plugin.bigquery;

import com.facebook.presto.spi.ColumnMetadata;
import com.google.cloud.bigquery.Field;
import com.google.cloud.bigquery.FieldList;
import com.google.common.collect.ImmutableMap;

import java.util.Collections;
import java.util.List;
import java.util.stream.Collectors;

import static com.google.cloud.bigquery.Field.Mode;
import static com.google.cloud.bigquery.Field.Mode.NULLABLE;
import static com.google.common.base.MoreObjects.firstNonNull;
import static com.google.common.collect.ImmutableMap.toImmutableMap;

public class Conversions
{
    private Conversions() {}

    static BigQueryColumnHandle toColumnHandle(Field field)
    {
        FieldList subFields = field.getSubFields();
        List<BigQueryColumnHandle> subColumns = subFields == null ?
                Collections.emptyList() :
                subFields.stream()
                        .map(Conversions::toColumnHandle)
                        .collect(Collectors.toList());
        return new BigQueryColumnHandle(
                field.getName(),
                BigQueryType.valueOf(field.getType().name()),
                getMode(field),
                subColumns,
                field.getDescription());
    }

    static ColumnMetadata toColumnMetadata(Field field)
    {
        return new ColumnMetadata(
                field.getName(), // name
                adapt(field).getPrestoType(),
                getMode(field) == NULLABLE, //nullable
                field.getDescription(), // comment
                null, // extraInfo
                false, // hidden
                ImmutableMap.of()); // properties
    }

    static BigQueryType.Adaptor adapt(Field field)
    {
        return new BigQueryType.Adaptor()
        {
            @Override
            public BigQueryType getBigQueryType()
            {
                return BigQueryType.valueOf(field.getType().name());
            }

            @Override
            public ImmutableMap<String, BigQueryType.Adaptor> getBigQuerySubTypes()
            {
                FieldList subFields = field.getSubFields();
                if (subFields == null) {
                    return ImmutableMap.of();
                }
                return subFields.stream().collect(toImmutableMap(Field::getName, Conversions::adapt));
            }

            @Override
            public Field.Mode getMode()
            {
                return Conversions.getMode(field);
            }
        };
    }

    private static Mode getMode(Field field)
    {
        return firstNonNull(field.getMode(), NULLABLE);
    }
}

File: velox/type/Conversions.cpp
/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "velox/type/Conversions.h"
#include "velox/functions/lib/string/StringImpl.h"

DEFINE_bool(
    experimental_enable_legacy_cast,
    false,
    "Experimental feature flag for backward compatibility with previous output"
    " format of type conversions used for casting. This is a temporary solution"
    " that aims to facilitate a seamless transition for users who rely on the"
    " legacy behavior and hence can change in the future.");

namespace facebook::velox::util {

/// folly's tryTo doesn't ignore control characters or other unicode whitespace.
/// We trim the string for control and unicode whitespace
/// from both directions and return a StringView of the result.
StringView trimWhiteSpace(const char* data, size_t length) {
  if (length == 0) {
    return StringView(data, 0);
  }

  int startIndex = 0;
  int endIndex = length - 1;
  const auto end = data + length;
  int size = 0;

  // We need to trim unicode chars and control chars
  // from left side of the string.
  for (auto i = 0; i < length;) {
    size = 0;
    auto isWhiteSpaceOrControlChar = false;

    if (data[i] & 0x80) {
      // Unicode - only check for whitespace.
      auto codePoint = utf8proc_codepoint(data + i, end, size);
      isWhiteSpaceOrControlChar =
          velox::functions::stringImpl::isUnicodeWhiteSpace(codePoint);
    } else {
      // Ascii - Check for both whitespace and control chars
      isWhiteSpaceOrControlChar =
          velox::functions::stringImpl::isAsciiWhiteSpace(data[i]) ||
          (data[i] > 0 && data[i] < 32);
    }

    if (!isWhiteSpaceOrControlChar) {
      startIndex = i;
      break;
    }

    i += size > 0 ? size : 1;
  }

  // Trim whitespace from right side.
  for (auto i = length - 1; i > startIndex;) {
    size = 0;
    auto isWhiteSpaceOrControlChar = false;

    if (data[i] & 0x80) {
      // Unicode - only check for whitespace.
      utf8proc_int32_t codePoint;
      // Find the right codepoint
      while ((codePoint = utf8proc_codepoint(data + i, end, size)) < 0 &&
             i > startIndex) {
        i--;
      }
      isWhiteSpaceOrControlChar =
          velox::functions::stringImpl::isUnicodeWhiteSpace(codePoint);
    } else {
      // Ascii - check if control char or whitespace
      isWhiteSpaceOrControlChar =
          velox::functions::stringImpl::isAsciiWhiteSpace(data[i]) ||
          (data[i] > 0 && data[i] < 32);
    }

    if (!isWhiteSpaceOrControlChar) {
      endIndex = i;
      break;
    }

    if (i > 0) {
      i--;
    }
  }

  // If we end on a unicode char make sure we add that to the end.
  auto charSize = size > 0 ? size : 1;
  return StringView(data + startIndex, endIndex - startIndex + charSize);
}

} // namespace facebook::velox::util
